### PR TITLE
Calculating the cpu ticks in nomad client

### PR DESCRIPTION
**Pull request by diptanu**

### CODE REVIEW

File: api/tasks.go [hunk 1]
```diff
@@ -19,6 +19,7 @@ type MemoryStats struct {
 type CpuStats struct {
 	SystemMode       float64
 	UserMode         float64
+	Ticks            float64
```
Comment (dadgar) [inline api/tasks.go]: Can we rename to `TotalTicks`

**Comment by dadgar** [comment]
Docker driver

**Comment by dadgar** [comment]
Can we do the same for host statistics too

File: client/driver/docker.go [hunk 2]
```diff
@@ -851,6 +856,9 @@ func (d *DockerDriver) Open(ctx *ExecContext, handleID string) (DriverHandle, er
 		doneCh:         make(chan bool),
 		waitCh:         make(chan *cstructs.WaitResult, 1),
 	}
+	if clkSpeed, err := shelpers.TotalTicksAvailable(); err == nil {
```
Comment (dadgar) [inline client/driver/docker.go]: This can just be part of the handle

File: client/driver/docker.go [hunk 3]
```diff
@@ -777,6 +779,9 @@ func (d *DockerDriver) Start(ctx *ExecContext, task *structs.Task) (DriverHandle
...
```
Comment (dadgar) [inline client/driver/docker.go]: Since you are checking in collectStats remove from here and open

File: client/driver/docker.go [hunk 4]
```diff
@@ -1015,6 +1023,13 @@ func (h *DockerHandle) collectStats() {
 					s.CPUStats.CPUUsage.UsageInUsermode, s.PreCPUStats.CPUUsage.UsageInUsermode,
 					s.CPUStats.CPUUsage.TotalUsage, s.PreCPUStats.CPUUsage.TotalUsage, cores)
 
+				if h.clkSpeed == 0.0 {
```
Comment (dadgar) [inline client/driver/docker.go]: create `h.getClockSpeed() float64` that does this check and stores it. That way the callers are clean

File: client/driver/executor/executor.go [hunk 5]
```diff
@@ -781,11 +789,13 @@ func (e *UniversalExecutor) aggregatedResourceUsage(pidStats map[string]*cstruct
 		totalSwap += pidStat.MemoryStats.Swap
 	}
 
+	totalTicks := e.systemCpuStats.TicksConsumed(percent)
 	totalCPU := &cstructs.CpuStats{
 		SystemMode: systemModeCPU,
 		UserMode:   userModeCPU,
 		Percent:    percent,
 		Measured:   ExecutorBasicMeasuredCpuStats,
+		TotalTicks: totalTicks,
```
Comment (dadgar) [inline client/driver/executor/executor.go]: Just put `e.systemCpuStats.TicksConsumed(percent)` right here. Not used elsewhere

File: client/driver/executor/executor.go [hunk 6]
```diff
@@ -710,7 +713,11 @@ func (e *UniversalExecutor) collectPids() {
 				e.logger.Printf("[DEBUG] executor: error collecting pids: %v", err)
 			}
 			e.pidLock.Lock()
-			e.pids = pids
+			for _, pid := range pids {
+				if _, ok := e.pids[pid.pid]; !ok {
+					e.pids[pid.pid] = pid
+				}
+			}
```
Comment (diptanu) [inline client/driver/executor/executor.go]: Need to prune here.

File: client/stats/cpu.go [hunk 7]
```diff
@@ -69,3 +60,13 @@ func (c *CpuStats) calculatePercent(t1, t2 float64, timeDelta int64) float64 {
 	overall_percent := (vDelta / float64(timeDelta)) * 100.0
 	return overall_percent
 }
+
+// getClkSpeed returns the total ticks available on a machine
+func (c *CpuStats) getClkSpeed() float64 {
```
Comment (dadgar) [inline client/stats/cpu.go]: Consolidate into a global helper. Talked offline

**Pull request merged**